### PR TITLE
[DevOps] Add improvements of YMLS Functional Tests

### PR DESCRIPTION
**Pull request by Batta32**
Related to MS3290

Depends on MS3505, MS3542, MS3601, MS3732, MS3736

### Purpose
*What is the context of this pull request? Why is it being done?*
The Functional Tests pipeline yamls for the C# and TS bots were improved to consume reusable templates.

### Changes
*Are there any changes that need to be called out as significant or particularly difficult to grasp? (Include illustrative screenshots for context if applicable.)*
We moved reusable functionality away from the main yamls `Nightly-Dotnet-VirtualAssistantToSkill.yml`, `Nightly-Dotnet-SkillSample.yml`, `Nightly-TypeScript-VirtualAssistantToSkill.yml`, and `Nightly-TypeScript-SkillSample.yml` and created with that the following templates added in the `yamls` folder on the root of the repository (which can be changed to another place)
- cleanupResourcesStep
- deletePackageLockStep
- deployStep
- dotnetBuildStep
- dotnetLogVersionsStep
- endLogStep
- functionalTestsStep
- nodeBuildStep
- nodeLogVersionsStep  
- npmInstallRequirementsStep
- setConfigFileStep
- setDirectLineStep
- useNodeStep
- useNugetStep
- yeomanInstallationStep
- yoCreateBotStep

image

### Tests
*Is this covered by existing tests or new ones? If no, why not?*

_Successful execution of Nightly-Dotnet-VirtualAssistantToSkill pipeline_
image

_Successful execution of Nightly-Dotnet-Skill pipeline_
image

_Successful execution of Nightly-TypeScript-VirtualAssistantToSkill pipeline_
image

_Successful execution of Nightly-TypeScript-Skill pipeline_
image

### Feature Plan
*Are there any remaining steps or dependencies before this issue can be fully resolved? If so, describe and link to any relevant pull requests or issues.*

#### General
- [X] I have commented my code, particularly in hard-to-understand areas	
- [ ] I have added or updated the appropriate tests	
- [X] I have updated related documentation

### CODE REVIEW

File: samples/csharp/assistants/virtual-assistant/VirtualAssistantSample.FunctionalTests/Nightly-Dotnet-VirtualAssistantToSkill.yml [hunk 1]
```diff
@@ -29,294 +28,68 @@ variables:
...
+
      botskills connect --remoteManifest "$skillManifestUrl" --luisFolder '..\..\..\skill\SkillSample\Deployment\Resources\LU\' --cs --languages 'en-us' --noRefresh --verbose
-    
+
      botskills refresh --cs --verbose
-    
 
     workingDirectory: 'samples/csharp/assistants/virtual-assistant/VirtualAssistantSample/'
     displayName: 'Test: VA - Connect SkillSample'
+    failOnStderr: false
```
Comment (matiasroldan6) [inline samples/csharp/assistants/virtual-assistant/VirtualAssistantSample.FunctionalTests/Nightly-Dotnet-VirtualAssistantToSkill.yml:92]: adding failOnStderr set to false seems superfluous, I think

File: templates/typescript/samples/sample-assistant/functional-tests/Nightly-TypeScript-VirtualAssistantToSkill.yml [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+     Write-Host $botWebAppName
+     Write-Host "`r`n"
+     echo "##vso[task.setvariable variable=BOTWEBAPPNAME;]$botWebAppName"
+
+     Write-Host "VA Resource Group"
+     Write-Host "===="
+     Write-Host $resourceGroupName
+     echo "##vso[task.setvariable variable=BOTRESOURCEGROUP;]$resourceGroupName"
+    workingDirectory: '$(WorkingDirectory)/src'
+    displayName: 'Deploy: VA - Get variables from appsettings'
```
Comment (matiasroldan6) [inline templates/typescript/samples/sample-assistant/functional-tests/Nightly-TypeScript-VirtualAssistantToSkill.yml:76]: In other yamls the name is 'Get _bot_ variables from appsettings'

Comment (Batta32) [reply templates/typescript/samples/sample-assistant/functional-tests/Nightly-TypeScript-VirtualAssistantToSkill.yml:76]: You can apply the changes, LGTM what you commented here

File: samples/csharp/assistants/virtual-assistant/VirtualAssistantSample.FunctionalTests/Nightly-Dotnet-VirtualAssistantToSkill.yml (hunk 1 above)
Comment (Batta32) [reply samples/csharp/assistants/virtual-assistant/VirtualAssistantSample.FunctionalTests/Nightly-Dotnet-VirtualAssistantToSkill.yml:92]: The default behavior is _false_?

Comment (matiasroldan6) [reply samples/csharp/assistants/virtual-assistant/VirtualAssistantSample.FunctionalTests/Nightly-Dotnet-VirtualAssistantToSkill.yml:92]: lemme double check but I'm 90% sure it is

File: templates/typescript/samples/sample-skill/functional-tests/Nightly-TypeScript-SkillSample.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  # ServicePrincipal: App Id of the Service Principal
+  # system.debug: System variable that can be set by the users. Set this to true to run the release in debug mode to assist in fault-finding.
+
+jobs:
+- job: Main
+  timeoutInMinutes: 0
+  variables:
+    BotAppId: $(AppId)
+    BotAppPassword: $(AppPassword)
+    BotType: "Skill"
```
Comment (matiasroldan6) [inline templates/typescript/samples/sample-skill/functional-tests/Nightly-TypeScript-SkillSample.yml:31]: Without the reference to yoCreateBotStep.yml, the variable BotType is unnecessary in all TS yamls

File: yaml/npmInstallRequirementsStep.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+steps:
+  - pwsh: |
+     # Install botdispatch and luis-apis from MyGet
+     npm install -g botdispatch luis-apis @latest --registry https://botbuilder.myget.org/F/botbuilder-tools-daily/npm/
```
Comment (matiasroldan6) [inline yaml/npmInstallRequirementsStep.yml:4]: do we need to install luis in the end?

Comment (Batta32) [reply yaml/npmInstallRequirementsStep.yml:4]: Yes

File: templates/typescript/samples/sample-skill/functional-tests/Nightly-TypeScript-SkillSample.yml (hunk 3 above)
Comment (Batta32) [reply templates/typescript/samples/sample-skill/functional-tests/Nightly-TypeScript-SkillSample.yml:31]: Ok, you can do the change if it's correct

File: yaml/endLogStep.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+#    Write-Host "Publish log"
+#    Write-Host "===="
+#    Get-Content -Path $(Join-Path "Deployment" "publish_log_after_skill.txt")
+#    Write-Host "`r`n"
+    failOnStderr: true
+    errorActionPreference: continue
+    workingDirectory: $(WorkingDirectory)
+    displayName: 'Debug: Show log contents'
+    #condition: always()
```
Comment (matiasroldan6) [inline yaml/endLogStep.yml:26]: is this right to be commented out? this is the step we wanted to always run, right?

Comment (Batta32) [reply yaml/endLogStep.yml:26]: I think that can be an improvement for the future. I don't think that we have time right now to apply it and validate it

Comment (matiasroldan6) [reply yaml/endLogStep.yml:26]: ok. And what about the commented out log ("publish_log_after_skill"). Can it be removed?

Comment (Batta32) [reply yaml/endLogStep.yml:26]: I'd prefer to don't change anything right now. Let's leave it as is unless if it's something that's making the yml fails

**Review by matiasroldan6 — APPROVED** [review]

**Pull request closed without merging**